torch/lib/c10d/Utils.hpp

```diff
@@ -0,0 +1,192 @@
+#pragma once
+
+#include <sys/socket.h>
+#include <sys/types.h>
+
+#include <cstdlib>
+#include <cstdint>
+#include <functional>
+#include <limits>
+#include <string>
+#include <system_error>
+#include <tuple>
+#include <vector>
+
+namespace c10d {
+
+
+using RankType = uint32_t;
+using PortType = uint16_t;
+using SizeType = uint64_t;
+
+#define SYSCHECK(expr) { \
+  errno = 0; auto ___output = (expr); (void)___output;     \
+  if (errno != 0) throw std::system_error(errno, std::system_category()); \
+}
+
+inline PortType convertToPort(int64_t port) {
+  if ((port < 0) || (port >= std::numeric_limits<PortType>::max()))
+    throw std::domain_error("invalid port (value out of range)");
+
+  return static_cast<PortType>(port);
+}
+
+inline RankType convertToRank(int64_t rank, int64_t min = 0) {
+  if ((rank < min) || (rank >= std::numeric_limits<RankType>::max()))
+    throw std::domain_error("invalid rank (value out of range)");
+
+  return static_cast<RankType>(rank);
+}
+
+// TCP util namespace
+namespace tcputil {
+
+
+// Send and receive
+template<typename T>
+void sendBytes(int socket,
+               const T* buffer,
+               size_t length,
+               bool moreData = false) {
+
+  size_t bytesToSend = sizeof(T) * length;
+  if (bytesToSend == 0) {
+    return;
+  }
+
+  auto bytes = reinterpret_cast<const uint8_t*>(buffer);
+  uint8_t* currentBytes = const_cast<uint8_t*>(bytes);
+
+  int flags = 0;
+
+#ifdef MSG_MORE
+  if (moreData) { // there is more data to send
+    flags |= MSG_MORE;
+  }
+#endif
+
+  while (bytesToSend > 0) {
+    ssize_t bytesSent;
+    SYSCHECK(bytesSent = ::send(socket, currentBytes, bytesToSend, flags))
+    if (bytesSent == 0) {
+      throw std::system_error(ECONNRESET, std::system_category());
+    }
+
+    bytesToSend -= bytesSent;
+    currentBytes += bytesSent;
+  }
+}
+
+template<typename T>
+void recvBytes(int socket, T* buffer, std::size_t length) {
+
+  size_t bytesToReceive = sizeof(T) * length;
+  if (bytesToReceive == 0) {
+    return;
+  }
+
+  auto bytes = reinterpret_cast<uint8_t*>(buffer);
+  uint8_t *currentBytes = bytes;
+
+  while (bytesToReceive > 0) {
+    ssize_t bytesReceived;
+    SYSCHECK(bytesReceived = ::recv(socket, currentBytes, bytesToReceive, 0))
+    if (bytesReceived == 0) {
+      throw std::system_error(ECONNRESET, std::system_category());
+    }
+
+    bytesToReceive -= bytesReceived;
+    currentBytes += bytesReceived;
+  }
```

Comment:
We did get some `read`/`write` wrappers in @pietern's PR. We might want to deduplicate those.

Reply:
I can't see the difference. Both are using `read`/`write` to receive or put data into file descriptors.

Reply:
they are different, one is using ::read, one is using ::recv. And we need to take care of the case when byteReceived == 0, we throw an error, that's not the logic in the File class's case

Reply:
What is the reason for send/recv vs read/write? Do we need the additional controls that they give us? No need to change this, but want to understand why not read/write.

Reply:
They should be the same. Used send/recv since the original THD util uses them.  The only difference is that here we check 0 bytes and throw an exception.